### PR TITLE
Rewrite integer lerp using intrinsics

### PR DESCRIPTION
**Pull request by vksnk**
Should be pretty much the same thing as before, but easier to read and maybe may produce better code.

### CODE REVIEW

File: src/Lerp.cpp [hunk 1]
```diff
@@ -134,17 +134,9 @@ Expr lower_lerp(Expr zero_val, Expr one_val, const Expr &weight) {
...
-
-                Expr rounding = Cast::make(UInt(2 * bits), 1) << Cast::make(UInt(2 * bits), (bits - 1));
-                Expr divisor = Cast::make(UInt(2 * bits), 1) << Cast::make(UInt(2 * bits), bits);
-
-                Expr prod_sum = zero_expand * inverse_typed_weight +
-                                one_expand * typed_weight + rounding;
-                Expr divided = ((prod_sum / divisor) + prod_sum) / divisor;
+                Expr shift = Cast::make(UInt(2 * bits), bits);
+                Expr prod_sum = widening_mul(zero_val, inverse_typed_weight) + widening_mul(one_val, typed_weight);
+                Expr divided = rounding_shift_right(rounding_shift_right(prod_sum, shift) + prod_sum, shift);
```
Comment (abadams) [inline src/Lerp.cpp:139]: This is equivalent to the original code, but I'm not convinced it's a good idea on x86. I bet you can do better with an average-round-up instruction.

**Comment by abadams** [comment]
I might spend some time trying to figure out how to do that division by 2^n - 1 more cheaply, but that would be a follow-up PR.

**Review by abadams — APPROVED** [review]

File: src/Lerp.cpp (hunk 1 above)
Comment (dsharletg) [inline src/Lerp.cpp:139]: This looks odd. I agree it should be the same as the old code, but I think the inner rounding_shift_right expression will be negligible. @zvookin @abadams any idea what this was about? Is it maybe attempting to ensure we actually hit one_val when typed_weight is the max?

Comment (abadams) [reply src/Lerp.cpp:139]: It's a trick to do x/255 as (x/256 + x)/256

Comment (dsharletg) [reply src/Lerp.cpp:139]: OK yeah, division by 2^n-1...

Comment (abadams) [reply src/Lerp.cpp:139]: Note that this is a special case of one of our unsigned division lowering methods. I was about to suggest just dividing and relying on that, but that only covers up to 255

Comment (abadams) [reply src/Lerp.cpp:139]: Actually I could easily extend that table to include 65535 and 2^32 - 1 ...

Comment (vksnk) [reply src/Lerp.cpp:139]: Ah, that's a cool trick, I've been wondering about that too.

Comment (abadams) [reply src/Lerp.cpp:139]: Hrm, no. This method has the benefit that it doesn't require additional widening.

Comment (abadams) [reply src/Lerp.cpp:139]: It's not equal! Testing it, I get the following identity instead:

`x/255 == (x/256 + x + 1) / 256`

Comment (abadams) [reply src/Lerp.cpp:139]: On x86, (x + 127)/255 (which is what I think it should be including rounding), is:

```
	vpaddw	(%rax,%rsi,2), %ymm0, %ymm2
	vpmulhuw	%ymm1, %ymm2, %ymm2
	vpsrlw	$7, %ymm2, %ymm2

```

and ((x + 127)/256 + x + 128) / 256, which is equal, gives:

```
	vpaddw	%ymm0, %ymm2, %ymm3
	vpsrlw	$8, %ymm3, %ymm3
	vpaddw	%ymm3, %ymm2, %ymm2
	vpaddw	%ymm1, %ymm2, %ymm2
	vpsrlw	$8, %ymm2, %ymm2

```

Comment (steven-johnson) [reply src/Lerp.cpp:139]: > It's a trick to do x/255 as (x/256 + x)/256

We should document that trick in a comment.

Comment (abadams) [reply src/Lerp.cpp:139]: So just change that code to something like:
```
Expr divisor = cast(UInt(2 * bits), (1ULL << bits) - 1);
Expr divided = (prod_sum + divisor/2) / divisor;
```

Comment (abadams) [reply src/Lerp.cpp:139]: Oh nuts, on arm it's reversed:
```
	add	v3.8h, v3.8h, v0.8h
	umull	v4.4s, v3.4h, v1.4h
	umull2	v3.4s, v3.8h, v2.8h
	ushr	v4.4s, v4.4s, #23
	ushr	v3.4s, v3.4s, #23
	xtn	v4.4h, v4.4s
	xtn2	v4.8h, v3.4s
	xtn	v3.8b, v4.8h

```

vs

```
	add	v3.8h, v2.8h, v0.8h
	usra	v2.8h, v3.8h, #8
	addhn	v2.8b, v2.8h, v1.8h
```

Comment (abadams) [reply src/Lerp.cpp:139]: btw the nasty thing about this trick is that it fails for the largest uint16_t, so you have to have some way to know that's not possible (as we do here).

Comment (abadams) [reply src/Lerp.cpp:139]: On the other hand, ```x/255 == ((x + 1)/256 + x + 1)/256``` seems to hold for all uint16 x.

So maybe the resolution here is teaching the arm backend this trick for division by 255

Comment (steven-johnson) [reply src/Lerp.cpp:139]: Do any of us have a copy of the book "Hacker's Delight"? IIRC it has quite a long section on this sort of technique, but I last read it at a different employer ~20 years ago...

**Comment by abadams** [comment]
My overall conclusion is that we should merge this PR with a comment explaining the trick and a TODO that notes it's actually one instruction cheaper to do the division directly on x86.

File: src/Lerp.cpp (hunk 1 above)
Comment (abadams) [reply src/Lerp.cpp:139]: Here's a useful identity for uint16s:

``` x / 255 === let y = saturating_sub(x, 127) in rounding_shift_right(rounding_shift_right(y, 8) + y, 8)```

On ARM we currently lower x/255 as:

```
	umull	v3.4s, v2.4h, v0.4h
	umull2	v2.4s, v2.8h, v1.8h
	ushr	v3.4s, v3.4s, #23
	ushr	v2.4s, v2.4s, #23
	xtn	v3.4h, v3.4s
	xtn2	v3.8h, v2.4s
	xtn	v2.8b, v3.8h
```

but using this identity we get:

```
	uqsub	v1.8h, v1.8h, v0.8h
	urshr	v2.8h, v1.8h, #8
	raddhn	v1.8b, v2.8h, v1.8h
```

Nice.

**Comment by steven-johnson** [comment]
Unrelated, but `correctness_custom_cuda_context` has been failing on Windows for a while now; is anyone looking into it?

**Comment by vksnk** [comment]
I am going to submit this (tests failure is unrelated) and create an issue to track if there are any better ways to do x / 255 (so discussion in the comments is not lost).

**Comment by abadams** [comment]
This actually hurt lerp on x86, because rounding shift right is more expensive than (x + 128) / 256.

Before:
```
	vpaddw	%ymm0, %ymm1, %ymm1
	vpsrlw	$8, %ymm1, %ymm2
	vpaddw	%ymm2, %ymm1, %ymm1
	vpsrlw	$8, %ymm1, %ymm1
	vextracti128	$1, %ymm1, %xmm2
	vpackuswb	%xmm2, %xmm1, %xmm1
```

After:
```
	vpsrlw	$7, %ymm2, %ymm3
	vpand	%ymm0, %ymm3, %ymm3
	vpsrlw	$8, %ymm2, %ymm4
	vpaddw	%ymm2, %ymm4, %ymm2
	vpaddw	%ymm3, %ymm2, %ymm2
	vpsrlw	$7, %ymm2, %ymm3
	vpand	%ymm0, %ymm3, %ymm3
	vpsrlw	$8, %ymm2, %ymm2
	vpaddw	%ymm2, %ymm3, %ymm2
	vpand	%ymm1, %ymm2, %ymm2
	vextracti128	$1, %ymm2, %xmm3
	vpackuswb	%xmm3, %xmm2, %xmm2
```

I'm changing the lowering on x86 anyway, but are there other targets that don't have a rounding_shift_right instruction?

File: src/Lerp.cpp (hunk 1 above)
Comment (alexreinking) [reply src/Lerp.cpp:141]: > Do any of us have a copy of the book "Hacker's Delight"? IIRC it has quite a long section on this sort of technique, but I last read it at a different employer ~20 years ago...

@steven-johnson - very late, I know, but I keep a copy on my desk. If it ever comes up again that we want to consult this book, let me know.

**Pull request merged**